[DAG] Remove OneUse restriction on sext when folding (shl (sext (add_nsw x, c1)), c2)

LiqinWeng opened the pull request:
This patch remove the restriction for folding (shl (sext (add_nsw x, c1)), c2) -> (add (shl (sext x), c2), c1 << c2), and test case from dhrystone , see this link: https://godbolt.org/z/8zfa3rnad

goldsteinn commented:
1. Can you split the new tests into a seperate commit s.t we can see the diff this patch generates.
2. "Remove restrictions" is a bit broad of a title for a patch only concerned with `(shl (sext (add_nsw x, c1)), c2)`. Maybe something along the lines of "Remove OneUse restriction on `sext` when folding `(shl (sext (add_nsw x, c1)), c2)`"?

goldsteinn commented on llvm/test/CodeGen/X86/pr65895.ll:
```diff
@@ -21,10 +21,11 @@ define i32 @PR65895() {
 ; CHECK-NEXT:    movsbl %al, %eax
 ; CHECK-NEXT:    movl %eax, d(%rip)
-; CHECK-NEXT:    leal 247(%rax,%rax,2), %eax
+; CHECK-NEXT:    leal 241(%rax,%rcx,2), %eax
```
This is an objective regression.

LiqinWeng replied:
fixed

RKSimon replied:
Regression still needs fixing

goldsteinn commented on llvm/test/CodeGen/X86/pr65895.ll:
```diff
@@ -21,10 +21,11 @@ define i32 @PR65895() {
 ; CHECK-NEXT:    # =>This Inner Loop Header: Depth=1
 ; CHECK-NEXT:    jmp .LBB0_2
 ; CHECK-NEXT:  .LBB0_3: # %for.end
+; CHECK-NEXT:    movzbl %al, %ecx
```
To address this can you add the OneUse checks back to the x86 version of `isDesirableToCommuteWithShift`?

LiqinWeng replied:
done

RKSimon commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -3304,6 +3304,13 @@ X86TargetLowering::preferredShiftLegalizationStrategy(
+bool X86TargetLowering::isDesirableToCommuteWithShift(
+    const SDNode *N, CombineLevel Level) const {
+  if (N->getOperand(0)->hasOneUse())
+    return true;
```
This doesn't look like the hasOneUse is being applied to the same operand as the original?

goldsteinn replied:
+1

topperc commented:
Isn't this only profitable for RISC-V if the addi can be absorbed into a load/store address. Couldn't there still be other tests like the X86 case that would regress?

RKSimon commented:
Are you intending to abandon this for #69105?

LiqinWeng commented:
> Are you intending to abandon this for #69105?

I found test of X86/pr65895.ll , llc -mtriple=riscv64 pr65895.ll , also regression(an extra register is used😂)；
this patch #69105， i will abandon；

LiqinWeng commented:
> Isn't this only profitable for RISC-V if the addi can be absorbed into a load/store address. Couldn't there still be other tests like the X86 case that would regress?

1. Test X86/pr65895.ll also regression（-mtriple=riscv64）
2. There is indeed no benefit in using x86 to test case riscv-shifted-extend.ll
So I don't know how to deal with it， should i remove the hasOneuse?

LiqinWeng committed:
[DAG] Remove OneUse restriction on sext when folding (shl (sext (add_nsw x, c1)), c2)

This patch remove the restriction for folding (shl (sext (add_nsw x, c1)), c2) -> (add (shl (sext x), c2), c1 << c2), and test case from dhrystone , see this link: https://godbolt.org/z/8zfa3rnad
